File: week5_dynamic_programming1/3_edit_distance/edit_distance.cpp
#include <iostream>
#include <string>

using std::string;

int minv(int a, int b, int c)
{
 int rv = a;
 if(b<a)
   rv = b;
 if(c<rv)
   rv = c;
 return rv;
}

int edit_distance(const string &str1, const string &str2) {
  //write your code here
  int m = str1.size();
  int n = str2.size();
  int ED[m+1][n+1];
  for(int i=0;i<m+1;i++)
    ED[i][0] = i;
  for(int j=0;j<n+1;j++)
    ED[0][j] = j;
  int diff = 1;
  for(int i=1;i<m+1;i++)
  for(int j=1;j<n+1;j++)
  {
      diff = 1;
      if(str1[i-1] == str2[j-1])
        diff = 0;
      ED[i][j] = minv(ED[i-1][j] +1 , ED[i][j-1]+1, ED[i-1][j-1] + diff);
   
  }
  return ED[m][n];
}

int main() {
  string str1;
  string str2;
  std::cin >> str1 >> str2;
  std::cout << edit_distance(str1, str2) << std::endl;
  return 0;
}

File: week5_dynamic_programming1/3_edit_distance/edit_distance.py
# Uses python3
def edit_distance(a, b):
    #write your code here
    T = [[float("inf")]*(len(b)+1) for _ in range(len(a)+1)]  
    for i in range(len(a)+1):
        T[i][0] = 0
    for j in range(len(b)+1):
        T[0][j] = 0
    for i in range(1,len(a)+1):
        for j in range(1,len(b)+1):
            diff = 0
            if (a[i-1] == b[j-1]):
                T[i][j] =  T[i-1][j-1] + 1
            else:
                T[i][j] = max(T[i][j-1],T[i-1][j])
            
            
    return T[len(a)][len(b)]

if __name__ == "__main__":
    print(edit_distance(input(), input()))
